castle/cms/_scripts/importjson.py

```diff
@@ -346,38 +346,38 @@ def import_object(filepath):
         obj.contentLayout = importtype.layout
         importtype.post_creation(obj)
         if not args.skip_transitioning and data['state']:
-            try:
-                print('Transitioning %s to %s' % (obj.id, data['state']))
-                api.content.transition(obj, to_state=data['state'])
-            except Exception:
-                logger.error("Error transitioning %s to %s, maybe workflows"
-                                " don't match up" % (obj.id, data['state']))
-                # pass
-                if stop_if_exception:
-                    if pdb_if_exception:
-                        pdb.set_trace()
-                    raise
+            # transition item only if it needs it
+            state = api.content.get_state(obj=obj)
+            if state != data['state']:
+                try:
+                    print('Transitioning %s to %s' % (obj.id, data['state']))
+                    api.content.transition(obj, to_state=data['state'])
+                except Exception:
+                    logger.error("Error transitioning %s to %s, maybe workflows"
+                                 " don't match up" % (obj.id, data['state']))
+                    # pass
+                    if stop_if_exception:
+                        if pdb_if_exception:
+                            pdb.set_trace()
+                        raise
 
         # set workflow / review history
         if 'review_history' in data:
             review_history = data['review_history']
             wtool = api.portal.get_tool(name='portal_workflow')
-            chain = wtool.getChainFor(obj)
-            if len(chain):
-                workflow_id = chain[0]
-                for h in review_history:
-                    wtool.setStatusOf(workflow_id, obj, h)
-                if len(chain) != 1:
-                    logger.warn('There should be 1 workflow for %s but'
-                                        'there are %i' % (path, len(chain)))
+            # loop over all workflow chains (usually only 1)
+            for workflow_id in wtool.getChainFor(obj):
+                obj.workflow_history[workflow_id] = review_history
```

Comment:
this is safe?

Reply:
getChainFor() returns the workflow ID as a string. Normally we would not have more than 1 chain. It seems to work fine for the normal case and fits with the code that was there before that was taking the first element of the list returned by getChainFor().

Reply:
I guess we'll never have more than one workflow assigned. If there are 2 assigned however, it seems to me like it'll set workflow history for each

Reply:
isn't that ok though? the import should result in content with matching metadata

Reply:
I remember Elro's presentation in DC about some financial instruments he created that had multiple workflows attached

Reply:
I've thought about it for OIE...